[Seah Shau Chung Nicholas] iP

nseah21 opened the pull request:
# Duke

> "Please use my app." - Nicholas

Duke *frees your mind* of having to remember things you need to do. It's, 

- text-based
- easy to learn
- ~~FAST~~ *SUPER FAST* to use 

All you need to do is,
1. download `duke.jar` from [here](https://github.com/nseah21/ip/releases/tag/A-Jar)
2. use the command line to navigate to the folder where it was saved 
3. run `java -jar duke.jar`
4. let it manage your tasks for you 😉 

And it is **FREE**! 🤑 

Features:
- [X] Managing tasks
- [x] Managing deadlines (coming soon)
- [x] Reminders (coming soon) 

If you are a Java programmer, you can use look at the source code to practice Java too. Here's the main method:

```java
public class Main {
    // ...
    public static void main(String[] args) {
        Duke duke = new Duke();
        duke.run();
    }
}
```

Commit:
Add Gradle support

nseah21 committed:
Add *.class files to .gitignore

nseah21 committed:
Add helper function for echoing commands

nseah21 committed:
Add scanner logic for echoing commands

nseah21 committed:
Add function for storing user input

nseah21 committed:
Add function for displaying user input history

nseah21 committed:
Add basic task class

nseah21 committed:
Remove unused imports

nseah21 committed:
Refactor Duke.java to work with task class

nseah21 committed:
Add logic for marking tasks as done/undone

nseah21 committed:
Add error handling for mark and unmark

nseah21 committed:
Rename task functions

nseah21 committed:
Add deadline class

nseah21 committed:
Add Todo class

nseah21 committed:
Add Event class

nseah21 committed:
Modify Task class to include toString()

nseah21 committed:
Add logic for tracking Todo, Event and Deadline tasks

nseah21 committed:
Modify input.txt and EXPECTED.TXT

nseah21 committed:
Add DukeException class

nseah21 committed:
Refactor main method and add error handling

nseah21 committed:
Make Task class abstract

nseah21 committed:
Refactor Duke.java with ArrayList instead of array

nseah21 committed:
Add error handling for event and deadline

nseah21 committed:
Add logic for deleting tasks

nseah21 committed:
Refactor task type from string to enum

nseah21 committed:
Change access modifier for enum

nseah21 committed:
Clean up code using switch statement

nseah21 committed:
Add Javadocs

nseah21 committed:
Add task data folder to .gitignore

nseah21 committed:
Add logic for saving tasks to hard disk

nseah21 committed:
Add logic for loading tasks into arraylist

nseah21 committed:
Refactor 'by' variable in Deadline to LocalDate

nseah21 committed:
Modify LocalDateTime implementation

nseah21 committed:
Add logic for using LocalDateTime

nseah21 committed:
Add regex for handling erroneous datetime

nseah21 committed:
Merge branch 'branch-Level-7'

nseah21 committed:
Merge branch 'branch-Level-8' and resolve conflicts

nseah21 committed:
Move UI functions into one class

nseah21 committed:
Move task list and functions into one class

nseah21 committed:
Move decision tree logic into one class

nseah21 committed:
Move save and load functions into one class

nseah21 committed:
Add function to return original 'by' String

nseah21 committed:
Refactor Duke.java to work with more OOP

nseah21 committed:
Change tasklist access to private

nseah21 committed:
Move classes into packages

nseah21 committed:
Merge branch 'master' into add-gradle-support

nseah21 committed:
Set up Gradle to build and run Duke

nseah21 committed:
Modify build.gradle

nseah21 committed:
Add test cases for unit testing

nseah21 committed:
Include JAR files from .gitignore

nseah21 committed:
Add JavaDocs

nseah21 committed:
Modify code to follow coding standard

nseah21 committed:
Add function for user to search for tasks

nseah21 committed:
Merge branch 'branch-A-JavaDoc'

nseah21 committed:
Merge branch 'branch-A-CodingStandard'

nseah21 committed:
Merge branch 'branch-Level-9'

nseah21 committed:
Merge branch 'add-gradle-support'

nseah21 committed:
Add Checkstyle with Gradle

nseah21 committed:
Update code based on checkstyle specs

janelleljt commented on src/main/java/entities/Task.java:
```diff
@@ -0,0 +1,72 @@
+package entities;
+
+/**
+ * A basic class for representing tasks
```
```suggestion
 * A basic class for representing tasks.
```

janelleljt reviewed (commented):
Looks good, cosmetic changes requested.

nseah21 committed:
Refactor utility classes for FXML

nseah21 committed:
Configure FXML for GUI

nseah21 committed:
Modify MainWindow.java

SpecOps2016 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,81 @@
+                this.taskList.listTasks();
+            } else {
+                try {
+                    Parser.decide(s, arr, this.taskList, this.storage);
```
decide is sounds a little bit vague to me. Perhaps a better verb can be used?

SpecOps2016 reviewed (commented):
Generally really well done, easy on the eyes with good verb usage for function names and variables with sensible nouns.

nseah21 committed:
Merge branch 'branch-A-Checkstyle'

nseah21 committed:
Merge branch 'branch-Level-10'

nseah21 committed:
Remove commented-out code

nseah21 committed:
Fix checkstyle errors

JustinPeng13 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,111 @@
+package duke;
```
Could all the packages here be put in a parent directory `duke` to show that they are associated, such that package names will be prefixed with `duke.`, e.g., `duke.duke`? I think it would be clearer.

JustinPeng13 commented on src/main/java/entities/Deadline.java:
```diff
@@ -0,0 +1,66 @@
+package entities;
```
Similarly, could this be placed in a `duke` directory such that the package is named `duke.entities`? Developers might not understand that `entities` package is specifically for the Duke application.

JustinPeng13 commented on src/main/java/utils/TaskList.java:
```diff
@@ -0,0 +1,180 @@
+     * @param remarks The remarks to be added for events or deadlines.
+     */
+    public void add(String description, Duke.TaskType type, String remarks) {
+        String s = "Got it. I've added this task:\n\t";
```
Could the variable `s` be renamed something clearer like `message`? Its scope is quite large for it to be a scratch variable.

JustinPeng13 reviewed (commented):
Good job overall! A few packages and variable names could be improved.

nseah21 committed:
Add assertions to utility classes

Exceptions are useful for detecting user errors, but can become
clunky when used to detect programmer errors.

Add assertions to the utility classes where the bulk of the
logic resides.

This will ensure that the programmer can check if his assumptions
are consistent with the existing code and make changes where
necessary.

nseah21 committed:
Refactor code using SLAP principles

Methods should not exceed 30 lines in length, as they
become difficult to read.

Abstract the code within each case statement into a
utility function within Parser.

This will clean up the code, and isolate the reader
from the nitty-gritty of how each case is being handled.

nseah21 committed:
Refactor code in Storage.java

nseah21 committed:
Use streams for searching tasks

Streams are lazy and generally incur better performance than manual
iteration. The code becomes more declarative and hence more
intuitive to the reader.

Replace the for-loop in TaskList.java with suitable stream operations.

This improves code quality and performance.

nseah21 committed:
Merge pull request #2 from nseah21/branch-A-Assertions

Merge branch 'branch-A-Assertions'

nseah21 committed:
Resolve merge conflicts

nseah21 committed:
Merge branch 'master' into branch-A-Streams

nseah21 committed:
Merge pull request #3 from nseah21/branch-A-CodeQuality

Merge branch 'branch-A-CodeQuality'

nseah21 committed:
Merge branch 'master' into branch-A-Streams

nseah21 committed:
Merge pull request #4 from nseah21/branch-A-Streams

Merge branch 'branch-A-Streams'

nseah21 committed:
Add extension for user to postpone deadlines

nseah21 committed:
Merge branch 'branch-BCD-Extension'

nseah21 committed:
Fix code quality issues

nseah21 committed:
Update assets

nseah21 committed:
Add background color to Duke response

nseah21 committed:
Fix incorrect filepath

nseah21 committed:
Add timeout for 'bye' command

nseah21 committed:
Add icon to GUI window

nseah21 committed:
Refactor listTasks method in TaskList

nseah21 committed:
Update README

nseah21 committed:
Add links to feature list in README

nseah21 committed:
Add bye command to README

nseah21 committed:
Update code using peer comments on PR

nseah21 committed:
Rename JAR file build

nseah21 committed:
Update README with instructions for MacOS

nseah21 committed:
Add personality to Ditto replies

nseah21 committed:
Update assets with new Ui.png

nseah21 committed:
Remove deep nesting from code

nseah21 committed:
Fix GUI issues and update Ui.png

nseah21 committed:
Add acknowledgements to README

nseah21 committed:
Give credit for code reuse

nseah21 committed:
Refactor Main and Launcher class

nseah21 committed:
Give credit for using image assets

nseah21 committed:
Flip GUI assets to face text
